docs/csharp/tutorials/attributes.md

```diff
@@ -0,0 +1,164 @@
+
+## How to restrict attribute usage
+
+Attributes can be used on a number of "targets". The above examples show them on classes, but they can also be used on:
+
+* Assembly
+* Class
+* Constructor
+* Delegate
+* Enum
+* Event
+* Field
+* GenericParameter
+* Interface
+* Method
+* Module
+* Parameter
+* Property
+* ReturnValue
+* Struct
+
+When you create an attribute class, by default, C# will allow you to use that attribute on any of the possible attribute targets. If you want to restrict your attribute
+to certain targets, you can do so by using the `AttributeUsageAttribute` on your attribute class. That's right, an attribute on an attribute!
+
+[!code-csharp[Using your own attribute](../../../samples/snippets/csharp/tutorials/attributes/Program.cs#AttributeUsageExample1)]
+
+If you attempt to put that attribute on something that's not a class or a struct, you will get a compiler error
+like `Attribute 'MyAttributeForClassAndStructOnly' is not valid on this declaration type. It is only valid on 'class, struct' declarations`
+
+[!code-csharp[Using your own attribute](../../../samples/snippets/csharp/tutorials/attributes/Program.cs#AttributeUsageExample2)]
+
+## How to use attributes attached to a code element
+
+Attributes act as metadata. Without some outward force, they won't actually do anything.
+
+To find and act on attributes, Reflection is generally needed. I won't cover Reflection in-depth in this tutorial, but the basic
+idea is that Reflection allows you to write code in C# that examines other code.
+
+For instance, you can use Reflection to get information about a class: 
+
+[!code-csharp[Getting type information with Reflection](../../../samples/snippets/csharp/tutorials/attributes/Program.cs#ReflectionExample1)]
+
+That will print out something like: `The assembly qualified name of MyClass is ConsoleApplication.MyClass, attributes, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null`
+
+Once you have a `TypeInfo` object (or a `MemberInfo`, `FieldInfo`, etc), you can use the `GetCustomAttributes` method. This will return a collection of `Attribute` objects.
+You can also use `GetCustomAttribute` and specify an Attribute type.
+
+Here's an example of using `GetCustomAttributes` on a `MemberInfo` instance for `MyClass` (which we saw earlier has an `[Obsolete]` attribute on it).
+
+[!code-csharp[Getting type information with Reflection](../../../samples/snippets/csharp/tutorials/attributes/Program.cs#ReflectionExample2)]
+
+That will print to console: `Attribute on MyClass: ObsoleteAttribute`. Try adding other attribute to `MyClass`.
+
+It's important to note that these `Attribute` objects are instantiated lazily. That is, they won't be instantiated until you use `GetCustomAttribute` or `GetCustomeAttributes`.
+They are also instantiated each time. Calling `GetCustomAttributes` twice in a row will return two different instances of `ObsoleteAttribute`.
+
+## What is the point of attributes?
```

Comment:
I'd like to see this as part of the beginning of the article. As I was reading in its current order, I felt this was heavy on the 'how' and very light on the 'why'. This is the start of that 'why' discussion.

This will be a great tutorial if this moves to the beginning, and gets expanded a bit more on why attributes are a useful concept.